compiler/plugins/target/AMD-AIE/iree-amd-aie/Target/AIETarget.cpp

```diff
@@ -356,6 +356,81 @@ FailureOr<std::vector<uint32_t>> loadUInt32ArrayFromFile(StringRef filePath) {
   return array;
 }
 
+struct Flatbuffer1dStringArrayConverter {
+  // The 1D array structure that represents the layout expected by the
+  // FlatBuffer schema.
+  SmallVector<std::string> data;
+  // Maps each ordinal to its corresponding index in the `data` array.
+  SmallVector<int32_t> indices;
+
+  Flatbuffer1dStringArrayConverter(uint64_t ordinalCount) {
+    indices.resize(ordinalCount);
+  }
+
+  void addEntry(uint64_t ordinal, StringRef entry) {
+    indices[ordinal] = data.size();
+    data.push_back(entry.str());
+  }
+
+  flatbuffers_string_vec_ref_t getFlatbufferVecRef(FlatbufferBuilder &builder) {
+    return builder.createStringVec(data);
+  }
+
+  template <typename FuncCreateStringRef>
+  SmallVector<flatbuffers_ref_t> getFlatbufferRefs(
+      FlatbufferBuilder &builder, FuncCreateStringRef createStringRef) {
+    SmallVector<flatbuffers_ref_t> dataRefs;
+    for (StringRef entry : data)
+      dataRefs.push_back(createStringRef(builder, builder.createString(entry)));
+    return dataRefs;
+  }
+};
+
+struct Flatbuffer3dUInt32ArrayConverter {
+  // The 3D array structure that represents the layout expected by the
+  // FlatBuffer schema.
+  SmallVector<SmallVector<std::vector<uint32_t>>> data;
+  // Maps an ordinal to the corresponding index in the outermost dimension of
+  // `data`. A value of -1 indicates that the given ordinal does not have a
+  // corresponding entry in `data`.
+  SmallVector<int32_t> indices;
+
+  Flatbuffer3dUInt32ArrayConverter(uint64_t ordinalCount) {
+    indices.resize(ordinalCount);
+  }
+
+  void addEntry(uint64_t ordinal,
+                const SmallVector<std::vector<uint32_t>> &entry2d) {
+    if (entry2d.empty()) {
+      // FlatBuffer does not support multidimensional arrays where any inner
+      // vector is empty, e.g., {{1, 2}, {}, {3, 4}} is invalid due to the empty
+      // second element. Handle this by setting the index to -1.
+      indices[ordinal] = -1;
+    } else {
+      indices[ordinal] = data.size();
+      data.push_back(entry2d);
+    }
+  }
+
+  template <typename Array1dRef, typename FuncCreateArray1d,
+            typename FuncCreateArray2d>
+  SmallVector<flatbuffers_ref_t> getFlatbufferRefs(
+      FlatbufferBuilder &builder, FuncCreateArray1d createArray1d,
+      FuncCreateArray2d createArray2d) {
+    SmallVector<flatbuffers_ref_t> dataRefs;
```

Comment:
Here, you could already reserve enough size: `dataRefs.reserve(data.size())` or use `llvm::map_to_vector`.